[UnifiedPDF] [iOS] Do not offer Writing Tools as a context menu action on text selections

aprotyas opened the pull request:
#### 58e49293f4769eeeed2a72d5694e5a080f8c02a2
<pre>
[UnifiedPDF] [iOS] Do not offer Writing Tools as a context menu action on text selections
<a href="https://bugs.webkit.org/show_bug.cgi?id=284782">https://bugs.webkit.org/show_bug.cgi?id=284782</a>
<a href="https://rdar.apple.com/141568926">rdar://141568926</a>

Reviewed by Aditya Keerthi.

The requisite WritingToolsController machinery for the Writing Tools
flow with text selections in the PDF plugin is not available yet. As
such, the Writing Tools context menu action is currently not usable.

This patch opts to disable the Writing Tools action - virtue of
WritingToolsBehaviorNone - when the active selection came from a plugin.
We will walk back this behavior once we fully support Writing Tools for
text selections in the PDF plugin, c.f. <a href="https://rdar.apple.com/141604977">rdar://141604977</a>.

To signal &quot;when the active selection came from a plugin&quot;, we make the
`isInPlugin` field of `EditorState` platform agnostic, and set the flag
to true whenever we have a non-null text selection in the PDF plugin.

Tests: WritingTools.PDFTextSelections

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::writingToolsBehavior const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::populateEditorStateIfNeeded const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleEditingCommand):
(WebKit::UnifiedPDFPlugin::isEditingCommandEnabled):
  Drive-by fix to make editing command handling case agnostic. Without
  this fix, we drop the `SelectAll` editing command on the floor when
  using `selectAll:` in the added test.

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, PDFTextSelections)):

Canonical link: <a href="https://commits.webkit.org/288032@main">https://commits.webkit.org/288032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77d7c060552b7bf3a0a404408a3e6842c67d52e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35671 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86261 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83823 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9063 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63756 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21478 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84787 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/925 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74367 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44042 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/824 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28541 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31166 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29143 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87700 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8957 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72094 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9142 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71323 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17759 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15407 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14323 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8908 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8749 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12272 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder reviewed (approved):
This seems testable 😅

aprotyas committed:
[UnifiedPDF] [iOS] Do not offer Writing Tools as a context menu action on text selections
https://bugs.webkit.org/show_bug.cgi?id=284782
rdar://141568926

Reviewed by Aditya Keerthi.

The requisite WritingToolsController machinery for the Writing Tools
flow with text selections in the PDF plugin is not available yet. As
such, the Writing Tools context menu action is currently not usable.

This patch opts to disable the Writing Tools action - virtue of
WritingToolsBehaviorNone - when the active selection came from a plugin.
We will walk back this behavior once we fully support Writing Tools for
text selections in the PDF plugin, c.f. rdar://141604977.

To signal "when the active selection came from a plugin", we make the
`isInPlugin` field of `EditorState` platform agnostic, and set the flag
to true whenever we have a non-null text selection in the PDF plugin.

Tests: WritingTools.PDFTextSelections

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator<<):
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::writingToolsBehavior const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::populateEditorStateIfNeeded const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleEditingCommand):
(WebKit::UnifiedPDFPlugin::isEditingCommandEnabled):
  Drive-by fix to make editing command handling case agnostic. Without
  this fix, we drop the `SelectAll` editing command on the floor when
  using `selectAll:` in the added test.

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, PDFTextSelections)):

Canonical link: https://commits.webkit.org/288032@main
